feat(perps): Add market-level allowlist/blocklist filtering for HIP-3

abretonc7s opened the pull request:
## **Description**

This PR implements market-level whitelisting and blacklisting for Perps HIP-3 markets, providing granular control over which perpetual trading markets are shown to users.

### What is the reason for the change?

Improvement from the initial remote feature flag implementation in PR #21823. While the initial PR established the remote feature flag infrastructure for Perps, we needed more granular control beyond simple DEX-level filtering. Specifically:

1. Users need the ability to enable/disable specific markets (e.g., "BTC", "xyz:XYZ100") rather than all markets from a DEX
2. Operators need both whitelist (enable only specific markets) and blacklist (block specific markets) capabilities
3. The system must support wildcard patterns for efficient bulk operations (e.g., "xyz:*" to enable/block all markets from a DEX)

### What is the improvement/solution?

**Migration from DEX-level to Market-level Filtering:**

- **Before**: Single `MM_PERPS_HIP3_ENABLED_DEXS` variable for whitelisting entire DEXs
- **After**: Two variables for granular market control:
  - `MM_PERPS_HIP3_ALLOWLIST_MARKETS` - Whitelist (empty = enable all, non-empty = only show these)
  - `MM_PERPS_HIP3_BLOCKLIST_MARKETS` - Blacklist (empty = block none, non-empty = hide these)

**Key Features:**

1. **Wildcard Pattern Support**
   - `"xyz:*"` or `"xyz"` - All markets from xyz DEX
   - `"xyz:TSLA"` - Specific market
   - `"BTC"` - Main DEX market

2. **Remote Feature Flag Integration**
   - LaunchDarkly provides remote configuration
   - Local env variables serve as fallback
   - "Sticky remote" pattern: once remote config loads, never downgrade to fallback

3. **Automatic Cache Invalidation & Reconnection**
   - `hip3ConfigVersion` increments when config changes
   - ConnectionManager monitors version via Redux
   - Automatically triggers cache clearing and provider reconnection
   - No manual intervention required

4. **Performance Optimizations**
   - Pattern caching (pre-compiled regex stored in Map)
   - Separate cache keys for filtered vs unfiltered data
   - `skipFilters` parameter for administrative queries

**Architecture:**

```
LaunchDarkly � RemoteFeatureFlagController � PerpsController
                                                   � (hip3ConfigVersion++)
                                           ConnectionManager
                                                   � (monitors version change)
                                           Clear caches + Reconnect
                                                   �
                                           HyperLiquidProvider
                                                   � (applies filters)
                                           Market list with patterns applied
```

## **Changelog**

CHANGELOG entry: Added market-level whitelisting and blacklisting for Perps HIP-3 markets with wildcard pattern support and automatic reconnection on configuration changes

## **Related issues**

Related to: #21823 (Initial remote feature flag implementation for Perps)

This PR splits out the market filtering functionality (Task #2) from the original feature flag work.

## **Manual testing steps**

```gherkin
Feature: Perps HIP-3 Market Filtering

  Scenario: Whitelist enables all markets when empty
    Given MM_PERPS_HIP3_ENABLED is true
    And MM_PERPS_HIP3_ALLOWLIST_MARKETS is empty
    And MM_PERPS_HIP3_BLOCKLIST_MARKETS is empty

    When user opens the Perps tab
    Then all available markets should be visible
    And markets from both main DEX and HIP-3 DEXs should appear

  Scenario: Whitelist restricts to specific markets
    Given MM_PERPS_HIP3_ENABLED is true
    And MM_PERPS_HIP3_ALLOWLIST_MARKETS is "BTC,ETH,xyz:TSLA"
    And MM_PERPS_HIP3_BLOCKLIST_MARKETS is empty

    When user opens the Perps tab
    Then only BTC, ETH, and xyz:TSLA markets should be visible
    And all other markets should be hidden

  Scenario: Whitelist with wildcard enables all markets from DEX
    Given MM_PERPS_HIP3_ENABLED is true
    And MM_PERPS_HIP3_ALLOWLIST_MARKETS is "xyz:*,BTC"
    And MM_PERPS_HIP3_BLOCKLIST_MARKETS is empty

    When user opens the Perps tab
    Then BTC market should be visible
    And all markets from xyz DEX should be visible
    And markets from other HIP-3 DEXs should be hidden

  Scenario: Blacklist blocks specific markets
    Given MM_PERPS_HIP3_ENABLED is true
    And MM_PERPS_HIP3_ALLOWLIST_MARKETS is empty
    And MM_PERPS_HIP3_BLOCKLIST_MARKETS is "xyz:TSLA,BTC"

    When user opens the Perps tab
    Then xyz:TSLA market should be hidden
    And BTC market should be hidden
    And all other markets should be visible

  Scenario: Blacklist with wildcard blocks all markets from DEX
    Given MM_PERPS_HIP3_ENABLED is true
    And MM_PERPS_HIP3_ALLOWLIST_MARKETS is empty
    And MM_PERPS_HIP3_BLOCKLIST_MARKETS is "xyz:*"

    When user opens the Perps tab
    Then all markets from xyz DEX should be hidden
    And markets from main DEX and other HIP-3 DEXs should be visible

  Scenario: Remote feature flag overrides local configuration
    Given local env has MM_PERPS_HIP3_ALLOWLIST_MARKETS="BTC"
    And LaunchDarkly is configured with perpsAllowlistMarkets="ETH,SOL"

    When app initializes and fetches remote feature flags
    Then only ETH and SOL markets should be visible
    And BTC market should be hidden (remote overrides local)

  Scenario: Automatic reconnection on configuration change
    Given user has Perps tab open with markets loaded
    And LaunchDarkly configuration is "BTC,ETH"

    When LaunchDarkly configuration changes to "SOL,AVAX"
    Then hip3ConfigVersion should increment
    And ConnectionManager should detect the change
    And app should automatically clear caches
    And app should reconnect to providers
    And only SOL and AVAX markets should be visible without manual refresh

  Scenario: Fallback to local configuration when remote unavailable
    Given LaunchDarkly is unreachable
    And local env has MM_PERPS_HIP3_ALLOWLIST_MARKETS="BTC,ETH"

    When user opens the Perps tab
    Then local configuration should be used as fallback
    And only BTC and ETH markets should be visible

  Scenario: Shorthand wildcard notation
    Given MM_PERPS_HIP3_ALLOWLIST_MARKETS is "xyz" (without :*)

    When user opens the Perps tab
    Then "xyz" should be interpreted as "xyz:*"
    And all markets from xyz DEX should be visible
```

## **Screenshots/Recordings**

### **Before**

N/A - This is a configuration-based feature without UI changes

### **After**

N/A - This is a configuration-based feature without UI changes

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I've included tests if applicable
- [ ] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

## Technical Implementation Details

### Environment Variables

**New variables added:**
```bash
# HIP-3 Feature Flags (remote override with local fallback)
export MM_PERPS_HIP3_ENABLED="true"
export MM_PERPS_HIP3_ALLOWLIST_MARKETS=""  # Whitelist: Empty = enable all markets
export MM_PERPS_HIP3_BLOCKLIST_MARKETS=""  # Blacklist: Empty = no blocking
```

**Removed variables:**
```bash
# OLD - Removed
export MM_PERPS_HIP3_ENABLED_DEXS=""  # Replaced by market-level filtering
export MM_PERPS_ENABLED_DEXS=""       # Replaced by market-level filtering
```

### Pattern Matching Examples

| Pattern | Matches | Description |
|---------|---------|-------------|
| `"BTC"` | `BTC` (main DEX) | Exact match on main DEX market |
| `"xyz:TSLA"` | `xyz:TSLA` | Exact match on HIP-3 market |
| `"xyz:*"` | `xyz:TSLA`, `xyz:AAPL`, etc. | All markets from xyz DEX |
| `"xyz"` | `xyz:TSLA`, `xyz:AAPL`, etc. | Shorthand for `"xyz:*"` |
| `""` (empty) | All markets | Discovery mode (whitelist) or no blocking (blacklist) |

### Pattern Matching Implementation

**Type Safety:**
- `CompiledPatternMatcher` - Type alias for pattern matchers (string for exact match, RegExp for wildcards)
- `CompiledPattern` - Interface combining original pattern with compiled matcher

**Compilation Strategy:**
```typescript
// Pattern → Compiled Matcher
"xyz:*"      → /^xyz:/         (prefix regex)
"xyz"        → /^xyz:/         (shorthand → prefix regex)
"xyz:TSLA"   → "xyz:TSLA"      (exact string match - fastest)
```

**Performance Optimization:**
1. All patterns pre-compiled at provider initialization via `recompileAllPatterns()`
2. Stored in typed arrays (`CompiledPattern[]`) for direct iteration
3. Eliminates repeated `compilePattern()` function call overhead during filtering
4. Better code clarity and type safety with pre-compiled matchers
5. Exact matches use string equality (fastest), wildcards use RegExp.test()

**Filtering Logic in `shouldIncludeMarket()`:**
```typescript
// Main DEX markets always included
if (dex === null) return true;

// Apply whitelist (if non-empty)
if (compiledEnabledPatterns.length > 0) {
  if (!compiledEnabledPatterns.some(p => matches(symbol, p.matcher))) {
    return false; // Not whitelisted
  }
}

// Apply blacklist (if non-empty)
if (compiledBlockedPatterns.length > 0) {
  if (compiledBlockedPatterns.some(p => matches(symbol, p.matcher))) {
    return false; // Blacklisted
  }
}

return true;
```

### New Selectors

- `selectHip3ConfigVersion()` - Returns version number for cache invalidation used by ConnectionManager to detect config changes

### Files Modified

**Core Logic (258 lines added):**
- `app/components/UI/Perps/controllers/PerpsController.ts`
  - `refreshHip3ConfigFromRemote()` - Extracts and validates remote config
  - Version tracking and increment on config change
  - "Sticky remote" pattern implementation

**Provider Implementation (411 lines added):**
- `app/components/UI/Perps/controllers/providers/HyperLiquidProvider.ts`
  - `CompiledPatternMatcher` type - Type alias for pattern matchers
  - `CompiledPattern` interface - Type for pre-compiled patterns
  - `recompileAllPatterns()` - Pre-compiles all patterns at initialization
  - `shouldIncludeMarket()` - Pattern matching logic using pre-compiled patterns
  - `matchesCompiledPattern()` - Fast pattern matching without Map lookups
  - `compilePattern()` - Regex compilation and caching
  - `skipFilters` parameter support
  - Separate cache keys for filtered/unfiltered data

**Selectors:**
- `app/components/UI/Perps/selectors/featureFlags/index.ts`
  - `selectHip3ConfigVersion` - Version selector for monitoring config changes

**Connection Management (22 lines added):**
- `app/components/UI/Perps/services/PerpsConnectionManager.ts`
  - Monitors `hip3ConfigVersion` changes via Redux
  - Triggers automatic cache clearing and reconnection

**Tests:**
- `app/components/UI/Perps/selectors/featureFlags/index.test.ts`
  - `selectHip3ConfigVersion` (3 tests) - Validates version tracking
  - Removed unused selector tests for unused allowlist/blocklist selectors

**Configuration:**
- `.js.env.example` - Environment variable documentation
- `bitrise.yml` - CI/CD environment configuration
- `app/core/Engine/controllers/perps-controller/index.ts` - Fallback parsing

**Types:**
- `app/components/UI/Perps/controllers/types/index.ts`
  - Updated `PerpsControllerConfig` interface
  - Added `skipFilters` to `GetMarketsParams`

### Performance Considerations

1. **Pattern Pre-Compilation**: All filter patterns are compiled once at initialization and stored in typed arrays (`CompiledPattern[]`), eliminating repeated `compilePattern()` function calls during market filtering
2. **Type-Safe Pattern Matchers**: Uses `CompiledPatternMatcher` type alias and `CompiledPattern` interface for better documentation and type safety
3. **Optimized Filtering**: `shouldIncludeMarket()` iterates pre-compiled arrays directly with compiled matchers readily available, improving code clarity and maintainability
4. **Separate Cache Keys**: Filtered and unfiltered data cached separately (`"${dex}_raw"` vs `"${dex}_filtered"`)
5. **StreamManager Integration**: Market metadata cached for 5 minutes, prices from WebSocket (real-time)

### Migration Guide

**From DEX-level to Market-level Filtering**

This PR replaces DEX-level filtering with more granular market-level filtering:

**Removed Environment Variables:**
```bash
# OLD - No longer used
export MM_PERPS_HIP3_ENABLED_DEXS="xyz,abc"  # Whitelist entire DEXs
export MM_PERPS_ENABLED_DEXS="xyz,abc"       # Alternative naming (removed)
```

**New Environment Variables:**
```bash
# NEW - Market-level control
export MM_PERPS_HIP3_ALLOWLIST_MARKETS=""  # Whitelist: empty = all markets (discovery mode)
export MM_PERPS_HIP3_BLOCKLIST_MARKETS=""  # Blacklist: empty = block nothing
```

**Common Migration Scenarios:**

1. **Enable all markets from specific DEXs (most common):**
   ```bash
   # Before: Enable all markets from xyz and abc DEXs
   MM_PERPS_HIP3_ENABLED_DEXS="xyz,abc"

   # After: Use wildcard patterns
   MM_PERPS_HIP3_ALLOWLIST_MARKETS="xyz:*,abc:*"
   # Or use shorthand (equivalent)
   MM_PERPS_HIP3_ALLOWLIST_MARKETS="xyz,abc"
   ```

2. **Enable specific markets only:**
   ```bash
   # Before: Not possible (DEX-level only)

   # After: List specific markets
   MM_PERPS_HIP3_ALLOWLIST_MARKETS="BTC,ETH,xyz:TSLA,xyz:AAPL"
   ```

3. **Enable all markets except specific ones:**
   ```bash
   # Before: Not possible (no blacklist support)

   # After: Use blacklist (whitelist empty = all markets)
   MM_PERPS_HIP3_ALLOWLIST_MARKETS=""
   MM_PERPS_HIP3_BLOCKLIST_MARKETS="xyz:SCAM,abc:RISKY"
   ```

4. **Block entire DEX:**
   ```bash
   # Before: Omit from ENABLED_DEXS list
   MM_PERPS_HIP3_ENABLED_DEXS="xyz"  # abc implicitly blocked

   # After: Use blacklist wildcard
   MM_PERPS_HIP3_ALLOWLIST_MARKETS=""     # Enable all
   MM_PERPS_HIP3_BLOCKLIST_MARKETS="abc:*"  # Block abc DEX
   # Or use shorthand
   MM_PERPS_HIP3_BLOCKLIST_MARKETS="abc"
   ```

5. **Enable all markets (discovery mode):**
   ```bash
   # Before: Leave ENABLED_DEXS empty or set to all known DEXs
   MM_PERPS_HIP3_ENABLED_DEXS=""

   # After: Leave ALLOWLIST_MARKETS empty
   MM_PERPS_HIP3_ALLOWLIST_MARKETS=""
   MM_PERPS_HIP3_BLOCKLIST_MARKETS=""
   ```

**LaunchDarkly Remote Feature Flags:**

The same patterns apply to LaunchDarkly configuration:
- `perpsAllowlistMarkets` - Array of market patterns (empty = all markets)
- `perpsBlocklistMarkets` - Array of market patterns to block (empty = block nothing)

**Example LaunchDarkly Config:**
```json
{
  "perpsHip3Enabled": true,
  "perpsAllowlistMarkets": ["BTC", "ETH", "xyz:*"],
  "perpsBlocklistMarkets": ["xyz:SCAM"]
}
```

### Breaking Changes

None - This is an additive change with backward-compatible fallback behavior.

**Note:** While `MM_PERPS_HIP3_ENABLED_DEXS` is no longer used, removing it from your configuration will not break anything. The new market-level filtering is more flexible and supersedes DEX-level filtering.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Implements market-level allowlist/blocklist filtering for HIP-3 perps with remote-flag overrides, caching, and automatic reconnection via hip3ConfigVersion.
> 
> - **Perps HIP-3 market filtering (allowlist/blocklist)**:
>   - Add pattern-based filtering (exact, wildcard, DEX shorthand) via `shouldIncludeMarket`, `compileMarketPattern`, etc. in `utils/marketUtils`.
>   - New parsing util `parseCommaSeparatedString` for LaunchDarkly/env values.
>   - New env vars in `.js.env.example` and CI (`bitrise.yml`): `MM_PERPS_HIP3_ENABLED`, `MM_PERPS_HIP3_ALLOWLIST_MARKETS`, `MM_PERPS_HIP3_BLOCKLIST_MARKETS` (replace old DEX-level flags).
> - **Controller & reconnection**:
>   - `PerpsController`: ingest remote flags for HIP-3 (enabled/allowlist/blocklist), maintain `hip3ConfigVersion`, and propagate config to provider.
>   - `PerpsConnectionManager`: monitor `selectHip3ConfigVersion` to clear caches and reconnect on config changes.
>   - Add `selectHip3ConfigVersion` selector; default state updated to include `hip3ConfigVersion`.
> - **Provider & subscriptions**:
>   - `HyperLiquidProvider`: apply market filtering, add market metadata caching (filtered/unfiltered), support `skipFilters`, and pass HIP-3 config to `HyperLiquidSubscriptionService`.
>   - `HyperLiquidSubscriptionService`: support HIP-3 on/off, map webData2 (main DEX) vs webData3 (multi-DEX), update feature flags without teardown; remove redundant clearinghouseState path.
> - **Types & tests**:
>   - Extend `GetMarketsParams` with `skipFilters`; update `PerpsControllerConfig` fallbacks.
>   - Extensive unit tests added/updated for HIP-3 config parsing, filtering, subscriptions, selectors, and connection manager.
> - **Misc**:
>   - Remove unused debug styles in `PerpsTabView.styles.ts`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d93ca772f511ce21f1e57ace950f2ffdc408bbbf. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

abretonc7s committed:
feat(perps): implement open interest cap

abretonc7s committed:
cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
revert pressable banner

abretonc7s committed:
feat: hip-3 subscriptions support

abretonc7s committed:
feat: cleanup

abretonc7s committed:
refactor: streamline client initialization in HyperLiquidProvider and enhance cache management in PerpsConnectionManager

- Simplified the client initialization process in HyperLiquidProvider by removing error handling and logging, ensuring the clientsInitialized flag is set only after successful initialization.
- Added cache clearing for oiCaps in PerpsConnectionManager to improve data management during reconnections.

abretonc7s committed:
dev warning

abretonc7s committed:
fix: unit tests

abretonc7s committed:
coverage

abretonc7s committed:
fix: add defensive validation for perpDexStates array bounds in HyperLiquidSubscriptionService

abretonc7s committed:
merge main into branch

abretonc7s committed:
fix: unit test coverage

abretonc7s committed:
fix: tpsl assignment

abretonc7s committed:
fix: improve error handling in PerpsOrderView and HyperLiquidProvider

- Updated PerpsOrderView to prevent error toast during initial market data load.
- Enhanced HyperLiquidProvider to avoid disabling DEX abstraction on network errors, ensuring the flag remains intact for future order verification.
- Added checks to handle transfer errors more gracefully, allowing for better management of DEX abstraction status.

abretonc7s committed:
fix: sonarcloud

abretonc7s committed:
feat: improve order validation

abretonc7s committed:
fix: unit tests

abretonc7s committed:
feat: implement open interest cap localization and refactor subscription parameters

- Updated PerpsOICapWarning component to use localized strings for open interest cap messages.
- Refactored subscribeToOICaps method signatures in PerpsController, HyperLiquidProvider, and HyperLiquidSubscriptionService to utilize a new SubscribeOICapsParams type for better clarity and maintainability.
- Added new localization entries for open interest cap reached and description in en.json.

abretonc7s committed:
chore: cleanup

abretonc7s committed:
cleanup

abretonc7s committed:
fix: bugbot

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/marketwhitelisting

abretonc7s committed:
feat: Implement HIP-3 feature flags for market filtering

This commit introduces new environment variables and selectors to support HIP-3 market filtering, allowing for enhanced control over which markets are enabled or blocked. Key changes include:

- Added environment variables for HIP-3 feature flags in `.js.env.example` and `bitrise.yml`.
- Updated `PerpsController` to manage HIP-3 configuration and versioning.
- Introduced selectors for retrieving enabled and blocked markets based on feature flags.
- Refactored `HyperLiquidProvider` and `HyperLiquidSubscriptionService` to utilize the new market filtering logic.

These changes aim to improve the flexibility and configurability of market access within the application.

abretonc7s committed:
merge main into branch

abretonc7s committed:
refactor: Enhance PerpsController and HyperLiquidProvider for improved market filtering

This commit refines the logging messages in the PerpsController for clarity and consistency. Additionally, it introduces a new pattern matching system in the HyperLiquidProvider to optimize market filtering performance. Key changes include:

- Updated log messages in PerpsController to remove redundant prefixes.
- Introduced `CompiledPattern` and `CompiledPatternMatcher` types for better pattern handling.
- Refactored pattern matching logic to utilize pre-compiled patterns, reducing lookup overhead.

These enhancements aim to streamline market filtering and improve code maintainability.

abretonc7s committed:
refactor: Improve market comparison logic in PerpsController

This commit enhances the market comparison logic in the PerpsController by ensuring that arrays are properly cloned before sorting and comparing. Key changes include:

- Updated the comparison of validated enabled and blocked markets to use spread syntax for array cloning.
- Improved readability and maintainability of the code.

These changes aim to prevent potential side effects from mutating the original arrays during comparison.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/marketwhitelisting

abretonc7s committed:
refactor: Enhance market filtering in HyperLiquidProvider

This commit refines the market filtering logic in the HyperLiquidProvider by introducing a new pattern matching system. Key changes include:

- Added `compileMarketPattern` and `shouldIncludeMarket` functions for optimized market filtering.
- Replaced previous pattern handling with pre-compiled patterns to improve performance.
- Updated tests to cover new pattern matching utilities.

These enhancements aim to streamline market filtering and improve code maintainability.

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/marketwhitelisting

abretonc7s committed:
refactor: Enhance DEX filtering logic in HyperLiquidSubscriptionService

This commit introduces improvements to the DEX filtering mechanism within the HyperLiquidSubscriptionService. Key changes include:

- Added a new `discoveredDexNames` property to maintain the order of DEXs for mapping.
- Implemented the `isDexEnabled` method to streamline the process of checking if a DEX is enabled based on configuration.
- Updated data processing logic to utilize the new filtering method, ensuring only relevant DEXs are processed during data callbacks.

These enhancements aim to improve the maintainability and performance of DEX data handling.

abretonc7s committed:
feat: Add debug view for feature flags in PerpsTabView

This commit introduces a debug container in the PerpsTabView component to display feature flag values during development. Key changes include:

- Added a new `debugContainer` style for the debug view.
- Integrated selectors for `selectPerpsEquityEnabledFlag`, `selectPerpsEnabledMarkets`, and `selectPerpsBlockedMarkets`.
- Implemented conditional rendering of the debug view to show feature flag states when in development mode.

These enhancements aim to facilitate easier debugging and validation of feature flags in the PerpsTabView.

michalconsensys commented on app/components/UI/Perps/controllers/providers/HyperLiquidProvider.ts:
```diff
@@ -319,7 +354,52 @@ export class HyperLiquidProvider implements IPerpsProvider {
+          availableHip3Dexs.push(dex.name);
+        }
+      });
+      return [null, ...availableHip3Dexs];
```
We could filter out the nulls from `availableHip3Dexs` with `isString` from `lodash` to make it easier to read

abretonc7s replied:
The availableHip3Dexs array is typed as string[] and only contains DEX names extracted from dex.name. The null
 check filters the source array allDexs, and we deliberately add null back in the return statement to represent the main DEX. There are no nulls in availableHip3Dexs to filter out, so lodash isString wouldn't apply here.

abretonc7s commented on app/components/UI/Perps/services/HyperLiquidSubscriptionService.ts:
```diff
@@ -703,25 +762,21 @@ export class HyperLiquidSubscriptionService {
+            // Only process DEXs we care about (skip others silently)
+            // webData3 API returns all protocol DEXs regardless of our config
+            if (!this.isDexEnabled(dexIdentifier)) {
+              return; // Skip this DEX - not enabled in our configuration
```
added guard

abretonc7s committed:
fix: bugbot

abretonc7s committed:
feat: Introduce parseCommaSeparatedString utility and refactor selectors

This commit adds a new utility function, `parseCommaSeparatedString`, to parse and clean up comma-separated strings into an array of trimmed, non-empty values. The function is utilized in the `selectPerpsEnabledMarkets` and `selectPerpsBlockedMarkets` selectors to simplify the parsing logic. Additionally, unit tests for the new utility function are included to ensure its correctness and robustness. These changes enhance code readability and maintainability while improving the handling of feature flag values.

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/marketwhitelisting

abretonc7s committed:
refactor: Remove debug features and clean up unused code in PerpsTabView and HyperLiquidSubscriptionService

This commit removes the debug container and associated feature flag selectors from the PerpsTabView component, enhancing code clarity. Additionally, it eliminates unused subscriptions and methods related to clearinghouse state in the HyperLiquidSubscriptionService, streamlining the service's functionality. These changes contribute to improved maintainability and performance of the codebase.

abretonc7s committed:
fix: linting

abretonc7s committed:
refactor: Simplify parsing of comma-separated strings in PerpsController

This commit refactors the parsing logic for comma-separated strings in the PerpsController by utilizing the newly introduced `parseCommaSeparatedString` utility. This change enhances code readability and maintainability. Additionally, it updates the maximum length check for market patterns in `marketUtils.ts` to prevent potential DoS attacks, increasing the limit from 100 to 200 characters.

abretonc7s committed:
pr comment

abretonc7s committed:
feat: Enhance market inclusion logic with HIP-3 master switch

This commit updates the `shouldIncludeMarket` function to incorporate a new parameter, `equityEnabled`, which acts as a master switch for HIP-3 markets. When `equityEnabled` is false, all HIP-3 markets are blocked. Additionally, the `HyperLiquidProvider` and `HyperLiquidSubscriptionService` are modified to utilize this new logic, ensuring that market inclusion checks are consistent across the application. Unit tests are also updated to reflect these changes, improving the overall robustness of market filtering.

abretonc7s committed:
fix: unit tests

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/marketwhitelisting

abretonc7s committed:
fix: unit test

Matt561 reviewed (commented):
Successfully tested the following cases on iOS simulator:
1. remote flag happy path: fallback -> remote flag upgrade with stickiness
2. Tiebreaker scenario where market is in allow and block list simultaneously
3. Blocking a market that a user has active position on -> Position is shown but market isn't discoverable

Found 2 bugs that may/may not be related to this PR but flagging for visibility:

1. Stock tab in market list doesn't filter out crypto perps markets

https://github.com/user-attachments/assets/d96a939c-b059-45e8-8229-55d5f4499dfc

2. Closing the "close all position" bottom sheet makes the app unresponsive. Looks like the bottom sheet doesn't fully close since we can still see the background opacity

https://github.com/user-attachments/assets/f3d14c16-f5d8-4b51-b5be-c15301b0546e

abretonc7s committed:
refactor: update HIP-3 market filtering to use allowlist and blocklist terminology

This commit refactors the PerpsController, HyperLiquidProvider, and related components to replace the terms "whitelist" and "blacklist" with "allowlist" and "blocklist" respectively. The changes include updates to environment variables, configuration options, and unit tests to reflect this new terminology. This enhances clarity and aligns with inclusive language practices across the codebase.

abretonc7s commented on app/components/UI/Perps/selectors/featureFlags/index.ts:
```diff
@@ -62,37 +64,87 @@ export const selectPerpsEquityEnabledFlag = createSelector(
+ * @returns string[] - Empty array = enable all markets (discovery mode), non-empty = whitelist
  */
-export const selectPerpsEnabledDexs = createSelector(
+export const selectPerpsEnabledMarkets = createSelector(
```
great point, I added them for debug but no need anymore!

abretonc7s committed:
fix: pr comments

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/marketwhitelisting

abretonc7s committed:
fix: invalid ff name

abretonc7s committed:
fix: unit test coverage

abretonc7s committed:
Merge branch 'main' into feat/perps/marketwhitelisting

abretonc7s committed:
fix: sonarcloud

abretonc7s committed:
feat: coverage

abretonc7s committed:
Merge remote-tracking branch 'origin/main' into feat/perps/marketwhitelisting

abretonc7s committed:
fix: even more coverage

weitingsun reviewed (approved):
mobile platform codeowner files LGTM

Matt561 commented on app/components/UI/Perps/selectors/featureFlags/index.ts:
```diff
@@ -62,37 +64,87 @@ export const selectPerpsEquityEnabledFlag = createSelector(
+ * @param state - Redux root state
+ * @returns number - Version increments when HIP-3 config changes
+ */
+export const selectHip3ConfigVersion = (state: RootState): number =>
```
No worries. Sounds good 👍
